// coffi/dist/index.js
// src/index.ts
import { dirname, join, resolve } from "node:path";

// src/logger.ts
class Logger {
  static instance;
  constructor() {}
  static getInstance() {
    if (!Logger.instance) {
      Logger.instance = new Logger;
    }
    return Logger.instance;
  }
  warn(message) {
    const yellowColor = "\x1B[33m";
    const resetColor = "\x1B[0m";
    console.warn(`${yellowColor}WARNING: ${message}${resetColor}`);
  }
  debug(message) {
    const blueColor = "\x1B[34m";
    const resetColor = "\x1B[0m";
    console.debug(`${blueColor}DEBUG: ${message}${resetColor}`);
  }
  error(message) {
    const redColor = "\x1B[31m";
    const resetColor = "\x1B[0m";
    console.error(`${redColor}ERROR: ${message}${resetColor}`);
  }
}
var logger = Logger.getInstance();

// src/utils.ts
import stripJsonComments from "strip-json-comments";
function cleanJson(json) {
  const stripped = stripJsonComments(json);
  return stripped.replace(/,\s*([}\]])/g, "$1").trim();
}

// src/index.ts
var DEFAULT_EXTENSIONS = [
  ".ts",
  ".mts",
  ".cts",
  ".js",
  ".mjs",
  ".cjs",
  ".json"
];
async function _exists(filepath) {
  try {
    return await Bun.file(filepath).exists();
  } catch {
    return false;
  }
}
async function parseConfigFile(filepath) {
  try {
    if (!await _exists(filepath)) {
      return null;
    }
    const ext = filepath.slice(filepath.lastIndexOf(".")).toLowerCase();
    if (ext === ".json") {
      const file = await Bun.file(filepath).text();
      const stripped = cleanJson(file);
      return JSON.parse(stripped);
    }
    const module = await import(filepath);
    let config = module.default || module;
    if (typeof config === "function") {
      config = config();
    }
    if (config instanceof Promise) {
      return config;
    }
    return config;
  } catch (error) {
    logger.error(`Failed to parse config file: ${parseErrorMessage(error)}`);
    return null;
  }
}
function parseErrorMessage(error) {
  if (error instanceof Error) {
    return error.message;
  }
  return String(error);
}
async function findPackageJson(cwd, maxDepth) {
  let currentDir = cwd;
  let currentDepth = 0;
  while (currentDepth < maxDepth) {
    const packageJsonPath = join(currentDir, "package.json");
    if (await _exists(packageJsonPath)) {
      return packageJsonPath;
    }
    const parentDir = dirname(currentDir);
    if (parentDir === currentDir) {
      break;
    }
    currentDir = parentDir;
    currentDepth++;
  }
  return null;
}
async function loadConfigInternal(name, extensions, cwd, maxDepth, preferredPath, packageJsonProperty) {
  if (packageJsonProperty) {
    const packageJsonPath = await findPackageJson(cwd, maxDepth);
    if (packageJsonPath) {
      const packageJson = await parseConfigFile(packageJsonPath);
      if (packageJson && typeof packageJson === "object" && packageJsonProperty in packageJson) {
        return {
          config: packageJson[packageJsonProperty],
          filepath: packageJsonPath
        };
      }
    }
  }
  if (preferredPath) {
    const resolvedPath = resolve(cwd, preferredPath);
    const config = await parseConfigFile(resolvedPath);
    if (config !== null) {
      return { config, filepath: resolvedPath };
    }
    logger.warn(`Preferred path "${preferredPath}" not found or invalid, searching for ${name} files instead.`);
  }
  let currentDir = cwd;
  let currentDepth = 0;
  while (currentDepth < maxDepth) {
    try {
      const files = await Array.fromAsync(new Bun.Glob("*").scan({ cwd: currentDir, onlyFiles: false }));
      const fileSet = new Set(files);
      for (const ext of extensions) {
        const filename = `${name}${ext}`;
        if (fileSet.has(filename)) {
          const filepath = join(currentDir, filename);
          const config = await parseConfigFile(filepath);
          if (config !== null) {
            return { config, filepath };
          }
        }
      }
    } catch (error) {}
    const parentDir = dirname(currentDir);
    if (parentDir === currentDir) {
      break;
    }
    currentDir = parentDir;
    currentDepth++;
  }
  return { config: null, filepath: null };
}
async function loadConfig(nameOrOptions, extensions = DEFAULT_EXTENSIONS, options = {}) {
  if (typeof nameOrOptions === "string") {
    const name2 = nameOrOptions;
    const {
      cwd: cwd2 = process.cwd(),
      maxDepth: maxDepth2 = 10,
      preferredPath: preferredPath2,
      packageJsonProperty: packageJsonProperty2
    } = options;
    return loadConfigInternal(name2, extensions, cwd2, maxDepth2, preferredPath2, packageJsonProperty2);
  }
  const {
    name,
    extensions: exts = DEFAULT_EXTENSIONS,
    cwd = process.cwd(),
    maxDepth = 10,
    preferredPath,
    packageJsonProperty
  } = nameOrOptions;
  return loadConfigInternal(name, exts, cwd, maxDepth, preferredPath, packageJsonProperty);
}
export {
  loadConfig
};
