// coffi/dist/index.cjs
var import_node_module = require("node:module");
var __create = Object.create;
var __getProtoOf = Object.getPrototypeOf;
var __defProp = Object.defineProperty;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __toESM = (mod, isNodeMode, target) => {
  target = mod != null ? __create(__getProtoOf(mod)) : {};
  const to = isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target;
  for (let key of __getOwnPropNames(mod))
    if (!__hasOwnProp.call(to, key))
      __defProp(to, key, {
        get: () => mod[key],
        enumerable: true
      });
  return to;
};
var __moduleCache = /* @__PURE__ */ new WeakMap;
var __toCommonJS = (from) => {
  var entry = __moduleCache.get(from), desc;
  if (entry)
    return entry;
  entry = __defProp({}, "__esModule", { value: true });
  if (from && typeof from === "object" || typeof from === "function")
    __getOwnPropNames(from).map((key) => !__hasOwnProp.call(entry, key) && __defProp(entry, key, {
      get: () => from[key],
      enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable
    }));
  __moduleCache.set(from, entry);
  return entry;
};
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, {
      get: all[name],
      enumerable: true,
      configurable: true,
      set: (newValue) => all[name] = () => newValue
    });
};

// src/index.ts
var exports_src = {};
__export(exports_src, {
  loadConfig: () => loadConfig
});
module.exports = __toCommonJS(exports_src);
var import_node_path = require("node:path");

// src/logger.ts
class Logger {
  static instance;
  constructor() {}
  static getInstance() {
    if (!Logger.instance) {
      Logger.instance = new Logger;
    }
    return Logger.instance;
  }
  warn(message) {
    const yellowColor = "\x1B[33m";
    const resetColor = "\x1B[0m";
    console.warn(`${yellowColor}WARNING: ${message}${resetColor}`);
  }
  debug(message) {
    const blueColor = "\x1B[34m";
    const resetColor = "\x1B[0m";
    console.debug(`${blueColor}DEBUG: ${message}${resetColor}`);
  }
  error(message) {
    const redColor = "\x1B[31m";
    const resetColor = "\x1B[0m";
    console.error(`${redColor}ERROR: ${message}${resetColor}`);
  }
}
var logger = Logger.getInstance();

// src/utils.ts
var import_strip_json_comments = __toESM(require("strip-json-comments"));
function cleanJson(json) {
  const stripped = import_strip_json_comments.default(json);
  return stripped.replace(/,\s*([}\]])/g, "$1").trim();
}

// src/index.ts
var DEFAULT_EXTENSIONS = [
  ".ts",
  ".mts",
  ".cts",
  ".js",
  ".mjs",
  ".cjs",
  ".json"
];
async function _exists(filepath) {
  try {
    return await Bun.file(filepath).exists();
  } catch {
    return false;
  }
}
async function parseConfigFile(filepath) {
  try {
    if (!await _exists(filepath)) {
      return null;
    }
    const ext = filepath.slice(filepath.lastIndexOf(".")).toLowerCase();
    if (ext === ".json") {
      const file = await Bun.file(filepath).text();
      const stripped = cleanJson(file);
      return JSON.parse(stripped);
    }
    const module2 = await import(filepath);
    let config = module2.default || module2;
    if (typeof config === "function") {
      config = config();
    }
    if (config instanceof Promise) {
      return config;
    }
    return config;
  } catch (error) {
    logger.error(`Failed to parse config file: ${parseErrorMessage(error)}`);
    return null;
  }
}
function parseErrorMessage(error) {
  if (error instanceof Error) {
    return error.message;
  }
  return String(error);
}
async function findPackageJson(cwd, maxDepth) {
  let currentDir = cwd;
  let currentDepth = 0;
  while (currentDepth < maxDepth) {
    const packageJsonPath = import_node_path.join(currentDir, "package.json");
    if (await _exists(packageJsonPath)) {
      return packageJsonPath;
    }
    const parentDir = import_node_path.dirname(currentDir);
    if (parentDir === currentDir) {
      break;
    }
    currentDir = parentDir;
    currentDepth++;
  }
  return null;
}
async function loadConfigInternal(name, extensions, cwd, maxDepth, preferredPath, packageJsonProperty) {
  if (packageJsonProperty) {
    const packageJsonPath = await findPackageJson(cwd, maxDepth);
    if (packageJsonPath) {
      const packageJson = await parseConfigFile(packageJsonPath);
      if (packageJson && typeof packageJson === "object" && packageJsonProperty in packageJson) {
        return {
          config: packageJson[packageJsonProperty],
          filepath: packageJsonPath
        };
      }
    }
  }
  if (preferredPath) {
    const resolvedPath = import_node_path.resolve(cwd, preferredPath);
    const config = await parseConfigFile(resolvedPath);
    if (config !== null) {
      return { config, filepath: resolvedPath };
    }
    logger.warn(`Preferred path "${preferredPath}" not found or invalid, searching for ${name} files instead.`);
  }
  let currentDir = cwd;
  let currentDepth = 0;
  while (currentDepth < maxDepth) {
    try {
      const files = await Array.fromAsync(new Bun.Glob("*").scan({ cwd: currentDir, onlyFiles: false }));
      const fileSet = new Set(files);
      for (const ext of extensions) {
        const filename = `${name}${ext}`;
        if (fileSet.has(filename)) {
          const filepath = import_node_path.join(currentDir, filename);
          const config = await parseConfigFile(filepath);
          if (config !== null) {
            return { config, filepath };
          }
        }
      }
    } catch (error) {}
    const parentDir = import_node_path.dirname(currentDir);
    if (parentDir === currentDir) {
      break;
    }
    currentDir = parentDir;
    currentDepth++;
  }
  return { config: null, filepath: null };
}
async function loadConfig(nameOrOptions, extensions = DEFAULT_EXTENSIONS, options = {}) {
  if (typeof nameOrOptions === "string") {
    const name2 = nameOrOptions;
    const {
      cwd: cwd2 = process.cwd(),
      maxDepth: maxDepth2 = 10,
      preferredPath: preferredPath2,
      packageJsonProperty: packageJsonProperty2
    } = options;
    return loadConfigInternal(name2, extensions, cwd2, maxDepth2, preferredPath2, packageJsonProperty2);
  }
  const {
    name,
    extensions: exts = DEFAULT_EXTENSIONS,
    cwd = process.cwd(),
    maxDepth = 10,
    preferredPath,
    packageJsonProperty
  } = nameOrOptions;
  return loadConfigInternal(name, exts, cwd, maxDepth, preferredPath, packageJsonProperty);
}
